### DISCUSSION TITLE
How does frb works under the hood?

### DISCUSSION THREAD

**Discussion by IamSanjid** [fzyzcjy/flutter_rust_bridge#1766] [General]
If this has been answered somewhere else I am sorry for the repitation but could not find anything useful with vanilla google search.
I am new to this flutter bridge type of projects but willing to try this out for future projects for it's features as it's one of the most popular "flutter bridge package".
So I really don't need very detailed explanation, just want to know how much abstraction layer there is.

How does it really call the Rust functions? My guess would be it compiles the rust code to native binary for the specific platform and uses Dart's `c-interoperability`? Something like [this](https://docs.flutter.dev/platform-integration/android/c-interop)?

Can I use platform specific API(s) directly in my rust code? Well, for this case it might not be easy but I am willing to go to the *sort of possible route*.

How does the Rust codebase calls Dart functions? Does it use some sort of messaging mechanism like `protobuf`?

Are there any potential limitations, besides the crate not being cross-platform, that I might encounter when adding external Rust crates?

**Comment by fzyzcjy** [comment]
Hi, firstly, https://cjycode.com/flutter_rust_bridge/guides/contributing/overview (the whole sections) introduces the internals. I guess it may answer some of your questions.

> Can I use platform specific API(s) directly in my rust code?

Sure!

> How does the Rust codebase calls Dart functions?

Dart provides mechanism to post to a (Dart) "port"

> Are there any potential limitations, besides the crate not being cross-platform, that I might encounter when adding external Rust crates?

I am not aware of any other limitations, since it works well on all 6 platforms flutter supports, and there are many tests running in the CI of this repo. But if you encounter any problem, feel free to create an issue!
